source/opt/const_folding_rules.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "const_folding_rules.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+const uint32_t kExtractCompositeIdInIdx = 0;
+
+ConstantFoldingRule FoldExtractWithConstants() {
+  return [](ir::Instruction* inst,
+            const std::vector<const analysis::Constant*>& constants)
+             -> const analysis::Constant* {
+    const analysis::Constant* c = constants[kExtractCompositeIdInIdx];
+    if (c == nullptr) {
+      return nullptr;
+    }
+
+    for (uint32_t i = 1; i < inst->NumInOperands(); ++i) {
+      uint32_t element_index = inst->GetSingleWordInOperand(i);
+      if (c->AsNullConstant()) {
+        // Return Null for the return type.
+        ir::IRContext* context = inst->context();
+        analysis::ConstantManager* const_mgr = context->get_constant_mgr();
+        analysis::TypeManager* type_mgr = context->get_type_mgr();
+        const analysis::NullConstant null_const(
+            type_mgr->GetType(inst->type_id()));
+        const analysis::Constant* real_const =
+            const_mgr->FindConstant(&null_const);
+        if (real_const == nullptr) {
+          ir::Instruction* const_inst =
+              const_mgr->GetDefiningInstruction(&null_const);
+          real_const = const_mgr->GetConstantFromInst(const_inst);
+        }
+        return real_const;
+      }
+
+      auto cc = c->AsCompositeConstant();
+      assert(cc != nullptr);
+      auto components = cc->GetComponents();
+      c = components[element_index];
```

Comment:
OpConstantComposite can have OpUndef as an element. What happens in this case?

Reply:
I forgot about that.  I think "c" ends up becoming nullptr.  I'll have to check that.

Reply:
The "constant" with the undef is not considered a constant by the constant manager.  So we hit the early exit at the top.

I remember an issue with this because, when specializing constants, we are not allowed to fold an OpSpecConstantComposite that contains an OpUndef.  I did not want to deal with that, so I left it as is.  At the time, I wanted to treat all OpUndef as if the were OpNull in the constant manager.

I've added a test case for this.